File: packages/dart_mod_common/lib/src/block_entity/flutter_display_block_entity.dart
/// Block entity for displaying Flutter UI content on block faces.
library;

import 'block_entity.dart';
import 'block_entity_settings.dart';
import '../types.dart';

/// Block entity for displaying Flutter UI content on block faces.
///
/// This block entity supports multi-block displays (e.g., 3x2 TV screen) by tracking
/// the grid position and total grid size. Each block in the grid references the same
/// Flutter surface but displays a different portion based on its grid coordinates.
///
/// ## Example - Single Block Display
///
/// ```dart
/// class TVBlock extends CustomBlock {
///   @override
///   BlockEntity? createBlockEntity(BlockPos pos, BlockState state) {
///     return FlutterDisplayBlockEntity(
///       surfaceId: 0,  // Use main Flutter surface
///       facing: Direction.north,
///       emissive: true,
///     );
///   }
/// }
/// ```
///
/// ## Example - Multi-Block Display (3x2 grid)
///
/// ```dart
/// // Each block in the 3x2 grid creates its own block entity with different grid positions
/// class BigTVBlock extends CustomBlock {
///   final int gridX;
///   final int gridY;
///
///   BigTVBlock({required this.gridX, required this.gridY});
///
///   @override
///   BlockEntity? createBlockEntity(BlockPos pos, BlockState state) {
///     return FlutterDisplayBlockEntity(
///       surfaceId: 0,
///       gridX: gridX,
///       gridY: gridY,
///       gridWidth: 3,
///       gridHeight: 2,
///       facing: Direction.north,
///     );
///   }
/// }
/// ```
class FlutterDisplayBlockEntity extends BlockEntity {
  /// The Flutter surface ID to display.
  ///
  /// - `0` = main Flutter surface (FlutterScreen)
  /// - `>0` = additional surfaces created via multi-surface API
  final int surfaceId;

  /// X position within a multi-block grid (0 = left).
  final int gridX;

  /// Y position within a multi-block grid (0 = top).
  final int gridY;

  /// Total grid width for multi-block displays.
  final int gridWidth;

  /// Total grid height for multi-block displays.
  final int gridHeight;

  /// Which block face displays the Flutter content.
  final Direction facing;

  /// Whether to use fullbright lighting (true = ignore world light).
  final bool emissive;

  /// Creates a Flutter display block entity.
  ///
  /// Parameters:
  /// - [surfaceId] The Flutter surface ID (0 for main surface)
  /// - [gridX] X position in multi-block grid (default 0)
  /// - [gridY] Y position in multi-block grid (default 0)
  /// - [gridWidth] Total grid width (default 1 for single block)
  /// - [gridHeight] Total grid height (default 1 for single block)
  /// - [facing] Which block face to render on (default north)
  /// - [emissive] Whether to use fullbright lighting (default true)
  FlutterDisplayBlockEntity({
    required String id,
    this.surfaceId = 0,
    this.gridX = 0,
    this.gridY = 0,
    this.gridWidth = 1,
    this.gridHeight = 1,
    this.facing = Direction.north,
    this.emissive = true,
  }) : super(
          settings: BlockEntitySettings(id: id),
        );

  @override
  void loadAdditional(Map<String, dynamic> nbt) {
    // Note: The actual properties are synced via Java NBT
    // This Dart-side is mainly for configuration
  }

  @override
  Map<String, dynamic> saveAdditional() {
    return {
      'SurfaceId': surfaceId,
      'GridX': gridX,
      'GridY': gridY,
      'GridWidth': gridWidth,
      'GridHeight': gridHeight,
      'Facing': facing.id,
      'Emissive': emissive,
    };
  }
}

File: packages/java_mc_bridge/src/main/java/com/redstone/blockentity/FlutterDisplayBlockEntity.java
package com.redstone.blockentity;

import com.redstone.DartBridge;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.core.HolderLookup;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.network.protocol.Packet;
import net.minecraft.network.protocol.game.ClientGamePacketListener;
import net.minecraft.network.protocol.game.ClientboundBlockEntityDataPacket;
import net.minecraft.world.level.block.entity.BlockEntity;
import net.minecraft.world.level.block.entity.BlockEntityType;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.level.storage.ValueInput;
import net.minecraft.world.level.storage.ValueOutput;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Block entity for displaying Flutter UI content on block faces.
 *
 * This block entity supports multi-block displays (e.g., 3x2 TV screen) by tracking
 * the grid position and total grid size. Each block in the grid references the same
 * Flutter surface but displays a different portion based on its grid coordinates.
 *
 * Key properties:
 * - surfaceId: Links to a Flutter surface (0 = main surface for now)
 * - gridX, gridY: Position within a multi-block grid (0,0 = top-left)
 * - gridWidth, gridHeight: Total grid dimensions
 * - facing: Which block face displays the content
 * - emissive: Whether to use fullbright lighting (true = ignore world light)
 */
public class FlutterDisplayBlockEntity extends BlockEntity {
    private static final Logger LOGGER = LoggerFactory.getLogger("FlutterDisplayBlockEntity");

    // Surface ID linking to Flutter - 0 means main surface
    private long surfaceId = 0;

    // Grid position for multi-block displays
    private int gridX = 0;
    private int gridY = 0;

    // Total grid dimensions (1x1 = single block display)
    private int gridWidth = 1;
    private int gridHeight = 1;

    // Which face of the block displays the content
    private Direction facing = Direction.NORTH;

    // Whether to use fullbright lighting (emissive)
    private boolean emissive = true;

    public FlutterDisplayBlockEntity(BlockEntityType<?> type, BlockPos pos, BlockState state) {
        super(type, pos, state);
    }

    // ==========================================================================
    // Getters and Setters
    // ==========================================================================

    public long getSurfaceId() {
        return surfaceId;
    }

    public void setSurfaceId(long surfaceId) {
        this.surfaceId = surfaceId;
        setChanged();
    }

    public int getGridX() {
        return gridX;
    }

    public void setGridX(int gridX) {
        this.gridX = gridX;
        setChanged();
    }

    public int getGridY() {
        return gridY;
    }

    public void setGridY(int gridY) {
        this.gridY = gridY;
        setChanged();
    }

    public int getGridWidth() {
        return gridWidth;
    }

    public void setGridWidth(int gridWidth) {
        this.gridWidth = Math.max(1, gridWidth);
        setChanged();
    }

    public int getGridHeight() {
        return gridHeight;
    }

    public void setGridHeight(int gridHeight) {
        this.gridHeight = Math.max(1, gridHeight);
        setChanged();
    }

    public Direction getFacing() {
        return facing;
    }

    public void setFacing(Direction facing) {
        this.facing = facing;
        setChanged();
    }

    public boolean isEmissive() {
        return emissive;
    }

    public void setEmissive(boolean emissive) {
        this.emissive = emissive;
        setChanged();
    }

    // ==========================================================================
    // Multi-block grid configuration
    // ==========================================================================

    /**
     * Configure this block entity for a multi-block display.
     *
     * @param surfaceId The Flutter surface ID to display
     * @param gridX X position in the grid (0 = left)
     * @param gridY Y position in the grid (0 = top)
     * @param gridWidth Total grid width
     * @param gridHeight Total grid height
     * @param facing Which block face to render on
     * @param emissive Whether to use fullbright lighting
     */
    public void configure(long surfaceId, int gridX, int gridY, int gridWidth, int gridHeight,
                         Direction facing, boolean emissive) {
        this.surfaceId = surfaceId;
        this.gridX = gridX;
        this.gridY = gridY;
        this.gridWidth = Math.max(1, gridWidth);
        this.gridHeight = Math.max(1, gridHeight);
        this.facing = facing;
        this.emissive = emissive;
        setChanged();

        // Sync to clients
        if (level != null && !level.isClientSide()) {
            level.sendBlockUpdated(getBlockPos(), getBlockState(), getBlockState(), 3);
        }
    }

    // ==========================================================================
    // NBT Persistence
    // ==========================================================================

    @Override
    protected void loadAdditional(ValueInput valueInput) {
        super.loadAdditional(valueInput);

        this.surfaceId = valueInput.getIntOr("SurfaceId", 0);
        this.gridX = valueInput.getIntOr("GridX", 0);
        this.gridY = valueInput.getIntOr("GridY", 0);
        this.gridWidth = valueInput.getIntOr("GridWidth", 1);
        this.gridHeight = valueInput.getIntOr("GridHeight", 1);

        int facingOrdinal = valueInput.getIntOr("Facing", Direction.NORTH.ordinal());
        this.facing = Direction.values()[facingOrdinal];

        this.emissive = valueInput.getBooleanOr("Emissive", true);
    }

    @Override
    protected void saveAdditional(ValueOutput valueOutput) {
        super.saveAdditional(valueOutput);

        valueOutput.putInt("SurfaceId", (int) surfaceId);
        valueOutput.putInt("GridX", gridX);
        valueOutput.putInt("GridY", gridY);
        valueOutput.putInt("GridWidth", gridWidth);
        valueOutput.putInt("GridHeight", gridHeight);
        valueOutput.putInt("Facing", facing.ordinal());
        valueOutput.putBoolean("Emissive", emissive);
    }

    // ==========================================================================
    // Client Sync
    // ==========================================================================

    @Override
    public CompoundTag getUpdateTag(HolderLookup.Provider registries) {
        CompoundTag tag = super.getUpdateTag(registries);
        tag.putLong("SurfaceId", surfaceId);
        tag.putInt("GridX", gridX);
        tag.putInt("GridY", gridY);
        tag.putInt("GridWidth", gridWidth);
        tag.putInt("GridHeight", gridHeight);
        tag.putInt("Facing", facing.ordinal());
        tag.putBoolean("Emissive", emissive);
        return tag;
    }

    @Override
    public Packet<ClientGamePacketListener> getUpdatePacket() {
        return ClientboundBlockEntityDataPacket.create(this);
    }
}
